db/table_cache.cc

```diff
@@ -487,6 +487,13 @@ Status TableCache::Get(
     if (s.ok()) {
       get_context->SetReplayLog(row_cache_entry);  // nullptr if no cache.
       s = t->Get(options, k, get_context, prefix_extractor.get(), skip_filters);
+      // If ts enabled for current cf, all cache entries will have ts appended
+      // at the end of their value.
+      if (s.ok() && row_cache_entry && !row_cache_entry->empty() &&
+          get_context->NeedTimestamp()) {
```

Comment:
This condition is checking the read request is trying to read back the timestamp, not that the column family enables timestamp. With this caching and retrieving logic, below cases could run into issues:

Assuming we have entry ("foo", 5, "v1")

First read: 
std::string timestamp;
Get(read_options.timestamp = 10, "foo", &timestamp); => this returns value "v1", and timestamp = 5, and stores ("foo"+10, "v1" + 5) entry in row cache.

Second read: Get(read_options.timestamp, "foo") => retrieves above entry from row cache, but because "get_context->NeedTimestamp()" is false because this get request is not reading back the timestamp. It will not try to remove the timestamp suffix from the row cache entry's value. So the result will be incorrect.

Similar issues for when first read that does the caching is not attempting to read timestamp and the second read that retrieves from caching is reading the timestamp.

To avoid this, storing timestamp in the replay log and read it back needs to be done unconditionally for when "ucmp_->timestamp_size() > 0". Also although it may store more than one timestamp, I think it's a better abstraction to isolate the timestamp logic  in `appendToReplayLog` and `replayGetContextLog`. With this, we can remove the hacky workaround to use `kTypeValue` to represent a timestamp's type.  I have a rough prototype for this approach here: https://github.com/facebook/rocksdb/compare/main...jowlyzhang:row_cache_return_ts let me know what you think.